To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

Read our **[Qwen3 Guide](https://docs.unsloth.ai/basics/qwen3-how-to-run-and-fine-tune)** and check out our new **[Dynamic 2.0](https://docs.unsloth.ai/basics/unsloth-dynamic-2.0-ggufs)** quants which outperforms other quantization methods!

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [2]:
from unsloth import FastLanguageModel
import torch
max_seq_length = 2048 # Choose any! We auto support RoPE Scaling internally!
dtype = None # None for auto detection. Float16 for Tesla T4, V100, Bfloat16 for Ampere+
load_in_4bit = True # Use 4bit quantization to reduce memory usage. Can be False.

# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/mistral-7b-bnb-4bit",
    "unsloth/mistral-7b-instruct-v0.2-bnb-4bit",
    "unsloth/llama-2-7b-bnb-4bit",
    "unsloth/llama-2-13b-bnb-4bit",
    "unsloth/codellama-34b-bnb-4bit",
    "unsloth/tinyllama-bnb-4bit",
    "unsloth/gemma-7b-bnb-4bit", # New Google 6 trillion tokens model 2.5x faster!
    "unsloth/gemma-2b-bnb-4bit",
] # More models at https://huggingface.co/unsloth

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "unsloth/mistral-7b-instruct-v0.3-bnb-4bit", # Choose ANY! eg teknium/OpenHermes-2.5-Mistral-7B
    max_seq_length = max_seq_length,
    dtype = dtype,
    load_in_4bit = load_in_4bit,
    # token = "hf_...", # use one if using gated models like meta-llama/Llama-2-7b-hf
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.4.7: Fast Mistral patching. Transformers: 4.51.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors:   0%|          | 0.00/4.14G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/157 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/141k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/587k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/446 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.96M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [3]:
model = FastLanguageModel.get_peft_model(
    model,
    r = 16, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj",],
    lora_alpha = 16,
    lora_dropout = 0, # Supports any, but = 0 is optimized
    bias = "none",    # Supports any, but = "none" is optimized
    # [NEW] "unsloth" uses 30% less VRAM, fits 2x larger batch sizes!
    use_gradient_checkpointing = "unsloth", # True or "unsloth" for very long context
    random_state = 3407,
    use_rslora = False,  # We support rank stabilized LoRA
    loftq_config = None, # And LoftQ
)

Unsloth 2025.4.7 patched 32 layers with 32 QKV layers, 32 O layers and 32 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the `ChatML` format for conversation style finetunes. We use [Open Assistant conversations](https://huggingface.co/datasets/philschmid/guanaco-sharegpt-style) in ShareGPT style. ChatML renders multi turn conversations like below:

```
<|im_start|>system
You are a helpful assistant.<|im_end|>
<|im_start|>user
What's the capital of France?<|im_end|>
<|im_start|>assistant
Paris.
```

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

We use our `get_chat_template` function to get the correct chat template. We support `zephyr, chatml, mistral, llama, alpaca, vicuna, vicuna_old` and our own optimized `unsloth` template.

Normally one has to train `<|im_start|>` and `<|im_end|>`. We instead map `<|im_end|>` to be the EOS token, and leave `<|im_start|>` as is. This requires no additional training of additional tokens.

Note ShareGPT uses `{"from": "human", "value" : "Hi"}` and not `{"role": "user", "content" : "Hi"}`, so we use `mapping` to map it.

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [5]:
import json
from xml.sax.saxutils import escape
from datasets import Dataset # To create a Hugging Face Dataset object

# --- XML Conversion Functions ---
def _json_to_xml_recursive(data, parent_key_name=None):
    """
    Recursively converts a Python dictionary or list to an XML string fragment.
    Handles basic types, nested dictionaries, and lists.
    Null values are represented as self-closing tags.
    """
    xml_parts = []
    if isinstance(data, dict):
        for key, value in data.items():
            safe_key = escape(str(key)) # Ensure key is a string and escaped
            if value is None:
                xml_parts.append(f"<{safe_key} />") # Self-closing tag for None
            elif isinstance(value, list):
                xml_parts.append(f"<{safe_key}>")
                for item in value:
                    item_tag = "item" # Default tag for list items
                    # Special handling for 'core_message_and_key_facts' list items
                    if key == "core_message_and_key_facts":
                        item_tag = "fact"
                    # Add other specific list item tag names here if needed for other array fields

                    xml_parts.append(f"<{escape(item_tag)}>")
                    if isinstance(item, (dict, list)): # Nested structure in list
                        xml_parts.append(_json_to_xml_recursive(item, item_tag))
                    else:
                        xml_parts.append(escape(str(item))) # Simple value in list
                    xml_parts.append(f"</{escape(item_tag)}>")
                xml_parts.append(f"</{safe_key}>")
            elif isinstance(value, dict): # Nested dictionary
                xml_parts.append(f"<{safe_key}>")
                xml_parts.append(_json_to_xml_recursive(value, key))
                xml_parts.append(f"</{safe_key}>")
            else: # Simple value
                xml_parts.append(f"<{safe_key}>{escape(str(value))}</{safe_key}>")
    elif isinstance(data, list): # If the data itself is a list (e.g., a list of objects not under a key)
        for item_idx, item in enumerate(data):
            item_tag = parent_key_name + "_item" if parent_key_name else f"item_{item_idx}"
            xml_parts.append(f"<{escape(item_tag)}>")
            if isinstance(item, (dict, list)):
                 xml_parts.append(_json_to_xml_recursive(item, item_tag))
            else:
                xml_parts.append(escape(str(item)))
            xml_parts.append(f"</{escape(item_tag)}>")

    return "".join(xml_parts)

def convert_json_to_xml(json_data_dict, root_name="extracted_data"):
    """
    Converts a Python dictionary (from JSON) to a complete XML string
    with a specified root element name.
    """
    if not isinstance(json_data_dict, dict):
        print(f"Warning: Input 'json_data_dict' for XML conversion was not a dictionary (type: {type(json_data_dict)}). Returning error XML.")
        return f"<{escape(root_name)}><error>Invalid or empty input data for XML conversion</error></{escape(root_name)}>"

    xml_content = _json_to_xml_recursive(json_data_dict)
    return f"<{escape(root_name)}>{xml_content}</{escape(root_name)}>"

# --- Data Loading and Transformation Function ---
def load_and_prepare_synthetic_dataset(jsonl_file_path="synthetic_outputs.jsonl"):
    """
    Loads data from a .jsonl file, removes specified metadata keys from
    'extracted_data', converts 'extracted_data' to XML,
    and formats it into the ShareGPT-style conversation list.
    """
    processed_conversations = []
    keys_to_remove = [
        "source_document_name",
        "extraction_timestamp", # Assuming this is metadata you want to remove from the final XML
        "original_base_json_input_name",
        "generation_method",
        "generation_timestamp",
        "source_document_length_chars" # This was in the original extracted_data but often not needed for the model input
    ]

    try:
        with open(jsonl_file_path, 'r', encoding='utf-8') as f:
            for i, line in enumerate(f):
                try:
                    if not line.strip():
                        print(f"Skipping empty line at index {i}.")
                        continue
                    data_entry = json.loads(line.strip())

                    extracted_data_json = data_entry.get("extracted_data")
                    output_markdown = data_entry.get("output_markdown_content")

                    if extracted_data_json is None:
                        print(f"Warning: 'extracted_data' is missing or null in line {i+1}. Creating error XML.")
                        xml_input = convert_json_to_xml({}, "extracted_data")
                    else:
                        # Make a copy to modify, to avoid altering the original dict if it's referenced elsewhere
                        modified_extracted_data = extracted_data_json.copy()
                        for key in keys_to_remove:
                            if key in modified_extracted_data:
                                del modified_extracted_data[key]
                        xml_input = convert_json_to_xml(modified_extracted_data, "extracted_data")

                    if output_markdown is None:
                        print(f"Warning: 'output_markdown_content' is missing or null in line {i+1}. Using empty string.")
                        output_markdown = ""

                    conversation = [
                        {"from": "human", "value": xml_input},
                        {"from": "gpt", "value": output_markdown}
                    ]
                    processed_conversations.append(conversation)

                except json.JSONDecodeError:
                    print(f"Skipping line {i+1} due to JSONDecodeError: {line.strip()[:100]}...")
                    continue
                except Exception as e:
                    print(f"Error processing line {i+1}: {e}. Line content: {line.strip()[:100]}...")
                    continue

    except FileNotFoundError:
        print(f"Error: The file '{jsonl_file_path}' was not found.")
        return Dataset.from_dict({"conversations": []})
    except Exception as e:
        print(f"An unexpected error occurred while trying to read '{jsonl_file_path}': {e}")
        return Dataset.from_dict({"conversations": []})

    if not processed_conversations:
        print("Warning: No valid conversations were processed. Returning an empty dataset.")

    return Dataset.from_dict({"conversations": processed_conversations})

# --- Example Usage (integrate this into your Colab notebook) ---

# 1. Make sure 'synthetic_outputs.jsonl' is accessible in your Colab environment.
jsonl_file = "synthetic_outputs.jsonl"

# 2. Load and prepare your synthetic dataset
print(f"Loading and preparing dataset from '{jsonl_file}'...")
synthetic_dataset = load_and_prepare_synthetic_dataset(jsonl_file)

print(f"\nSuccessfully loaded and processed {len(synthetic_dataset)} entries.")

if len(synthetic_dataset) > 0:
    print("\nExample of one processed entry (ShareGPT style, before applying chat template):")
    print(json.dumps(synthetic_dataset[0]['conversations'], indent=2))

    # ... (rest of your tokenizer setup and dataset mapping code) ...

else:
    print("\nNo data was loaded. Please check your 'synthetic_outputs.jsonl' file and path.")


Loading and preparing dataset from 'synthetic_outputs.jsonl'...

Successfully loaded and processed 4129 entries.

Example of one processed entry (ShareGPT style, before applying chat template):
[
  {
    "from": "human",
    "value": "<extracted_data><document_type>Internal Memo Snippet</document_type><primary_subject_or_headline_focus>Media strategy for Sarah Lin joining QuantumVest's Board to drive APAC fintech expansion</primary_subject_or_headline_focus><publication_date_or_dateline>2019-03-15</publication_date_or_dateline><key_entities_and_roles><primary_organization><name>QuantumVest</name><role_in_document>subject_of_media_strategy</role_in_document></primary_organization></key_entities_and_roles><core_message_and_key_facts><fact>The campaign aims to elevate QuantumVest's profile among Asia-Pacific fintech stakeholders.</fact><fact>A press announcement about Sarah Lin's appointment to the Board is scheduled for 8am Hong Kong time on March 15, 2019.</fact><fact>Prior to release, 

Let's see how the `ChatML` format works by printing the 5th element

In [ ]:
dataset[5]["conversations"]

[{'from': 'human',
  'value': 'What is the typical wattage of bulb in a lightbox?'},
 {'from': 'gpt',
  'value': 'The typical wattage of a bulb in a lightbox is 60 watts, although domestic LED bulbs are normally much lower than 60 watts, as they produce the same or greater lumens for less wattage than alternatives. A 60-watt Equivalent LED bulb can be calculated using the 7:1 ratio, which divides 60 watts by 7 to get roughly 9 watts.'},
 {'from': 'human',
  'value': 'Rewrite your description of the typical wattage of a bulb in a lightbox to only include the key points in a list format.'}]

In [ ]:
print(dataset[5]["text"])

<|im_start|>user
What is the typical wattage of bulb in a lightbox?<|im_end|>
<|im_start|>assistant
The typical wattage of a bulb in a lightbox is 60 watts, although domestic LED bulbs are normally much lower than 60 watts, as they produce the same or greater lumens for less wattage than alternatives. A 60-watt Equivalent LED bulb can be calculated using the 7:1 ratio, which divides 60 watts by 7 to get roughly 9 watts.<|im_end|>
<|im_start|>user
Rewrite your description of the typical wattage of a bulb in a lightbox to only include the key points in a list format.<|im_end|>



If you're looking to make your own chat template, that also is possible! You must use the Jinja templating regime. We provide our own stripped down version of the `Unsloth template` which we find to be more efficient, and leverages ChatML, Zephyr and Alpaca styles.

More info on chat templates on [our wiki page!](https://github.com/unslothai/unsloth/wiki#chat-templates)

In [18]:
from unsloth.chat_templates import get_chat_template

# Assume 'tokenizer' is already loaded and initialized, for example:
# from transformers import AutoTokenizer
# tokenizer = AutoTokenizer.from_pretrained("your-base-model-name")

# 1. Define your custom template and EOS token
unsloth_template = \
    "{{ bos_token }}"\
    "{{ 'You are a helpful assistant to the user\\n' }}"\
    "{% for message in messages %}"\
        "{% if message['role'] == 'user' %}"\
            "{{ '>>> User: ' + message['content'] + '\\n' }}"\
        "{% elif message['role'] == 'assistant' %}"\
            "{{ '>>> Assistant: ' + message['content'] + eos_token + '\\n' }}"\
        "{% endif %}"\
    "{% endfor %}"\
    "{% if add_generation_prompt %}"\
        "{{ '>>> Assistant: ' }}"\
    "{% endif %}"

unsloth_eos_token = "eos_token" # Or "</s>" if that's what your model/Unsloth expects for its true EOS

# 2. Apply the chat template to your tokenizer
# The 'if False:' is removed so this code executes.
tokenizer = get_chat_template(
    tokenizer,
    chat_template = (unsloth_template, unsloth_eos_token,), # Provide as a tuple
    mapping = {"role" : "from", "content" : "value", "user" : "human", "assistant" : "gpt"}, # For ShareGPT style
    map_eos_token = True, # Important for Unsloth to handle EOS correctly, often maps to </s>
)

def formatting_prompts_func(examples):
  # 'examples' will be a batch from your Hugging Face dataset.
  # examples["conversations"] will be a list of conversations.
  # Each 'convo' in 'convos' is a list of turns (dictionaries).
  convos = examples["conversations"]

  texts = [tokenizer.apply_chat_template(convo, tokenize=False, add_generation_prompt=False) for convo in convos]

  # Returns a dictionary mapping the new column name "text" to the list of processed strings.
  return {"text": texts}

print("Formatting dataset...")
# Apply the formatting function to create the "text" column
# Ensure synthetic_dataset is not empty before mapping
if len(synthetic_dataset) > 0:
    dataset_ready_for_trainer = synthetic_dataset.map(
        formatting_prompts_func,
        batched=True,  # Process data in batches for efficiency
        # num_proc=2,  # Optional: Use multiple processes if your dataset is large
        remove_columns=["conversations"] # Optional: Remove the original column if no longer needed
    )
    print("Dataset formatted.")
    print(f"First example text: {dataset_ready_for_trainer[0]['text']}")
else:
    print("Synthetic dataset is empty. Cannot format.")
    # Handle this case appropriately, perhaps by exiting or raising an error
    # For now, creating an empty shell to avoid subsequent errors if trainer is still called.
    from datasets import Dataset
    dataset_ready_for_trainer = Dataset.from_dict({"text":[]})

# Now your 'tokenizer' is updated with the custom chat template.
# You can proceed to use it with your dataset and the formatting_prompts_func:

# Example (assuming synthetic_dataset is loaded as per previous steps):
# def formatting_prompts_func(examples):
#     convos = examples["conversations"]
#     texts = [tokenizer.apply_chat_template(convo, tokenize = False, add_generation_prompt = False) for convo in convos]
#     return { "text" : texts, }
#
# if len(synthetic_dataset) > 0:
#     try:
#         dataset_formatted = synthetic_dataset.map(formatting_prompts_func, batched = True)
#         print("\nExample of one formatted text entry (after applying custom chat template):")
#         print(dataset_formatted[0]['text'])
#     except Exception as e:
#         print(f"\nAn error occurred during the .map() operation: {e}")
# else:
#     print("\nDataset is empty, cannot apply formatting.")


Formatting dataset...


Map:   0%|          | 0/4129 [00:00<?, ? examples/s]

Dataset formatted.
First example text: <s>You are a helpful assistant to the user
>>> User: <extracted_data><document_type>Internal Memo Snippet</document_type><primary_subject_or_headline_focus>Media strategy for Sarah Lin joining QuantumVest's Board to drive APAC fintech expansion</primary_subject_or_headline_focus><publication_date_or_dateline>2019-03-15</publication_date_or_dateline><key_entities_and_roles><primary_organization><name>QuantumVest</name><role_in_document>subject_of_media_strategy</role_in_document></primary_organization></key_entities_and_roles><core_message_and_key_facts><fact>The campaign aims to elevate QuantumVest's profile among Asia-Pacific fintech stakeholders.</fact><fact>A press announcement about Sarah Lin's appointment to the Board is scheduled for 8am Hong Kong time on March 15, 2019.</fact><fact>Prior to release, an embargoed version will be shared with Nikkei Asia for potential feature coverage and interview with Sarah Lin.</fact><fact>If Nikkei Asia pa

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [19]:
from trl import SFTTrainer
from transformers import TrainingArguments
from unsloth import is_bfloat16_supported
# Ensure 'model', 'tokenizer', 'max_seq_length' are defined
# Use 'dataset_ready_for_trainer' which now has the "text" column
if len(dataset_ready_for_trainer) > 0 : # Only proceed if dataset has content
    trainer = SFTTrainer(
        model=model,
        tokenizer=tokenizer,
        train_dataset=dataset_ready_for_trainer, # Pass the pre-processed dataset
        # formatting_func=None, # REMOVE or set to None
        dataset_text_field="text",      # Tell SFTTrainer to use the "text" column directly
        max_seq_length=max_seq_length,
        dataset_num_proc=2,
        packing=False,
        args=TrainingArguments(
            per_device_train_batch_size=2,
            gradient_accumulation_steps=4,
            warmup_steps=5,
            max_steps=60,
            learning_rate=2e-4,
            fp16=not is_bfloat16_supported(),
            bf16=is_bfloat16_supported(),
            logging_steps=1,
            optim="adamw_8bit",
            weight_decay=0.01,
            lr_scheduler_type="linear",
            seed=3407,
            output_dir="outputs",
            report_to="none",
        ),
    )
    print("SFTTrainer initialized successfully with pre-formatted dataset.")
else:
    print("Cannot initialize SFTTrainer: Formatted dataset is empty.")


Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/4129 [00:00<?, ? examples/s]

SFTTrainer initialized successfully with pre-formatted dataset.


In [ ]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.748 GB.
4.52 GB of memory reserved.


In [20]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 4,129 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 41,943,040/7,000,000,000 (0.60% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [ ]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

870.2993 seconds used for training.
14.5 minutes used for training.
Peak reserved memory = 6.289 GB.
Peak reserved memory for training = 1.769 GB.
Peak reserved memory % of max memory = 42.643 %.
Peak reserved memory for training % of max memory = 11.995 %.


<a name="Inference"></a>
### Inference
Let's run the model! Since we're using `ChatML`, use `apply_chat_template` with `add_generation_prompt` set to `True` for inference.

In [ ]:
from unsloth.chat_templates import get_chat_template

tokenizer = get_chat_template(
    tokenizer,
    chat_template = "chatml", # Supports zephyr, chatml, mistral, llama, alpaca, vicuna, vicuna_old, unsloth
    mapping = {"role" : "from", "content" : "value", "user" : "human", "assistant" : "gpt"}, # ShareGPT style
    map_eos_token = True, # Maps <|im_end|> to </s> instead
)

FastLanguageModel.for_inference(model) # Enable native 2x faster inference

messages = [
    {"from": "human", "value": "Continue the fibonnaci sequence: 1, 1, 2, 3, 5, 8,"},
]
inputs = tokenizer.apply_chat_template(
    messages,
    tokenize = True,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
).to("cuda")

outputs = model.generate(input_ids = inputs, max_new_tokens = 64, use_cache = True)
tokenizer.batch_decode(outputs)

Unsloth: Will map <|im_end|> to EOS = <|im_end|>.
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


['<|im_start|>user\nContinue the fibonnaci sequence: 1, 1, 2, 3, 5, 8,<|im_end|> \n<|im_start|>assistant\nThe next number in the Fibonacci sequence is 13.\n\nThe Fibonacci sequence is a series of numbers in which each number is the sum of the two preceding ones, starting from 0 and 1. The sequence goes: 0, 1, 1, ']

 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [ ]:
FastLanguageModel.for_inference(model) # Enable native 2x faster inference

messages = [
    {"from": "human", "value": "Continue the fibonnaci sequence: 1, 1, 2, 3, 5, 8,"},
]
inputs = tokenizer.apply_chat_template(
    messages,
    tokenize = True,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
).to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(input_ids = inputs, streamer = text_streamer, max_new_tokens = 128, use_cache = True)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


<|im_start|>user
Continue the fibonnaci sequence: 1, 1, 2, 3, 5, 8,<|im_end|> 
<|im_start|>assistant
The next number in the Fibonacci sequence is 13.

The Fibonacci sequence is a series of numbers in which each number is the sum of the two preceding ones, starting from 0 and 1. The sequence goes: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [ ]:
model.save_pretrained("lora_model")  # Local saving
tokenizer.save_pretrained("lora_model")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/tokenizer.json')

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

In [ ]:
if False:
    from unsloth import FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name = "lora_model", # YOUR MODEL YOU USED FOR TRAINING
        max_seq_length = max_seq_length,
        dtype = dtype,
        load_in_4bit = load_in_4bit,
    )
    FastLanguageModel.for_inference(model) # Enable native 2x faster inference

messages = [
    {"from": "human", "value": "What is a famous tall tower in Paris?"},
]
inputs = tokenizer.apply_chat_template(
    messages,
    tokenize = True,
    add_generation_prompt = True, # Must add for generation
    return_tensors = "pt",
).to("cuda")

from transformers import TextStreamer
text_streamer = TextStreamer(tokenizer)
_ = model.generate(input_ids = inputs, streamer = text_streamer, max_new_tokens = 128, use_cache = True)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


<|im_start|>user
What is a famous tall tower in Paris?<|im_end|> 
<|im_start|>assistant
The Eiffel Tower is a famous tall tower in Paris. It is one of the most recognizable landmarks in the world and is a popular tourist destination. The tower was built in 1889 for the World's Fair and is named after Gustave Eiffel, the engineer who designed and built it. The tower stands at a height of 324 meters (1,063 feet) and is made of iron. It is located on the Champ de Mars, a large public park in Paris.<|im_end|>


You can also use Hugging Face's `AutoModelForPeftCausalLM`. Only use this if you do not have `unsloth` installed. It can be hopelessly slow, since `4bit` model downloading is not supported, and Unsloth's **inference is 2x faster**.

In [ ]:
if False:
    # I highly do NOT suggest - use Unsloth if possible
    from peft import AutoModelForPeftCausalLM
    from transformers import AutoTokenizer

    model = AutoModelForPeftCausalLM.from_pretrained(
        "lora_model",  # YOUR MODEL YOU USED FOR TRAINING
        load_in_4bit=load_in_4bit,
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

In [22]:
# Save to 8bit Q8_0
if True: model.save_pretrained_gguf("model", tokenizer,)
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "q4_k_m", token = "")

Unsloth: ##### The current model auto adds a BOS token.
Unsloth: ##### Your chat template has a BOS token. We shall remove it temporarily.


Unsloth: Merging 4bit and LoRA weights to 16bit...
Unsloth: Will use up to 4.47 out of 12.67 RAM for saving.
Unsloth: Saving model... This might take 5 minutes ...


100%|██████████| 32/32 [01:45<00:00,  3.29s/it]


Unsloth: Saving tokenizer... Done.
Unsloth: Saving model/pytorch_model-00001-of-00003.bin...
Unsloth: Saving model/pytorch_model-00002-of-00003.bin...
Unsloth: Saving model/pytorch_model-00003-of-00003.bin...
Done.
==((====))==  Unsloth: Conversion from QLoRA to GGUF information
   \\   /|    [0] Installing llama.cpp might take 3 minutes.
O^O/ \_/ \    [1] Converting HF to GGUF 16bits might take 3 minutes.
\        /    [2] Converting GGUF 16bits to ['q8_0'] might take 10 minutes each.
 "-____-"     In total, you will have to wait at least 16 minutes.

Unsloth: Installing llama.cpp. This might take 3 minutes...
Unsloth: [1] Converting model at model into q8_0 GGUF format.
The output location will be /content/model/unsloth.Q8_0.gguf
This might take 3 minutes...
INFO:hf-to-gguf:Loading model: model
INFO:hf-to-gguf:Model architecture: MistralForCausalLM
INFO:gguf.gguf_writer:gguf: This GGUF file is for Little Endian only
INFO:hf-to-gguf:Exporting model...
INFO:hf-to-gguf:gguf: loading mod

Unsloth: ##### The current model auto adds a BOS token.
Unsloth: ##### We removed it in GGUF's chat template for you.


Unsloth: Conversion completed! Output location: /content/model/unsloth.Q8_0.gguf


Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
